actors/miner/tests/util.rs

```diff
@@ -2279,9 +2284,341 @@ fn fixed_hasher(offset: ChainEpoch) -> Box<dyn Fn(&[u8]) -> [u8; 32]> {
+                    &format!("no clean-up epoch for pre-commit at {}", precommit.pre_commit_epoch),
+                );
+                precommit_total += &precommit.pre_commit_deposit;
+                Ok(())
+            });
+            acc.require_no_error(ret, "error iterating pre-commited sectors");
+        }
+        Err(e) => {
+            acc.add(&format!("error loading precommited_sectors: {e}"));
+        }
+    };
+
+    acc.require(state.pre_commit_deposits == precommit_total,&format!("sum of pre-commit deposits {precommit_total} does not equal recorded pre-commit deposit {}", state.pre_commit_deposits));
+}
+
 #[allow(dead_code)]
-pub fn check_state_invariants(_rt: &MockRuntime) {
-    // TODO check state invariants
+pub fn check_state_invariants(rt: &MockRuntime) -> (StateSummary, MessageAccumulator) {
+    let state: State = rt.get_state();
+    let store = rt.store();
+    let balance = rt.balance.clone().take();
+    let acc = MessageAccumulator::default();
+    let sector_size;
+
+    let mut miner_summary =
+        StateSummary { deadline_cron_active: state.deadline_cron_active, ..Default::default() };
+
+    // load data from linked structures
+    match state.get_info(store) {
+        Ok(info) => {
+            miner_summary.window_post_proof_type = info.window_post_proof_type;
+            sector_size = info.sector_size;
+            check_miner_info(info, &acc);
+        }
+        Err(e) => {
+            // Stop here, it's too hard to make other useful checks.
+            acc.add(&format!("error loading miner info: {e}"));
+            return (miner_summary, acc);
+        }
+    };
+
+    check_miner_balances(rt.policy(), &state, store, &balance, &acc);
+
+    let allocated_sectors = match store.get_cbor::<BitField>(&state.allocated_sectors) {
+        Ok(Some(allocated_sectors)) => {
+            if let Some(sectors) = allocated_sectors.bounded_iter(1 << 30) {
+                sectors.map(|i| i as SectorNumber).collect()
+            } else {
+                acc.add(&format!("error expanding allocated sector bitfield"));
+                BTreeSet::new()
+            }
+        }
+        Ok(None) => {
+            acc.add(&format!("error loading allocated sector bitfield"));
+            BTreeSet::new()
+        }
+        Err(e) => {
+            acc.add(&format!("error loading allocated sector bitfield: {e}"));
+            BTreeSet::new()
+        }
+    };
+
+    check_precommits(rt.policy(), &state, store, &allocated_sectors, &acc);
+
+    let mut all_sectors: BTreeMap<SectorNumber, SectorOnChainInfo> = BTreeMap::new();
+    match Sectors::load(&store, &state.sectors) {
+        Ok(sectors) => {
+            let ret = sectors.amt.for_each(|sector_number, sector| {
+                all_sectors.insert(sector_number, sector.clone());
+                acc.require(
+                    allocated_sectors.is_empty() || allocated_sectors.contains(&sector_number),
```

Comment:
Remove the is_empty half of the conditional.  The nil check in go is attempting to not panic during failed allocatedSectorsMap load.  There is a case to be made for repeating that approach here because it makes a failure to load allocated sectors numbers bitfield very noisy.  However the no-failure-on-empty approach also causes error suppression in the case of validly parsed empty sector numbers bitfield and non empty sectors array.  I'd prefer extra noisy complete failing than the possibility of error suppression.

Reply:
makes sense, thanks!